tests/upgrades/scenario_workers.py

```diff
@@ -0,0 +1,58 @@
+import json
+import os
+
+import pytest
+from automation_tools.satellite6.hammer import set_hammer_config
+from fabric.api import env
+
+import robottelo
+from robottelo.config import settings
+
+
+_json_file = 'upgrade_workers.json'
+
+
+def save_worker_id(test_name, worker_id):
+    data = {}
+    if os.path.exists(_json_file):
+        with open(_json_file) as rjson:
+            data = json.load(rjson)
+    with open(_json_file, 'w') as wjson:
+        # Removing the parameter name from test name before save
+        test_name = test_name.split('[')[0] if '[' in test_name else test_name
+        data.update({test_name: worker_id})
+        wjson.seek(0)
+        json.dump(data, wjson)
+
+
+def get_worker_id(test_name):
+    with open(_json_file) as rjson:
+        data = json.load(rjson)
+        return data.get(test_name)
+
+
+@pytest.fixture(autouse=True)
+def save_pretest_worker_id(request, worker_id):
+    """Saves the worker id of pre_upgrade test in json"""
+    if request.node.get_closest_marker('pre_upgrade'):
+        save_worker_id(request.node.name, worker_id)
+
+
+@pytest.fixture(autouse=True)
+def set_posttest_worker_id(request):
+    """Retrieves worker_id of pre_upgrade and sets the same to run post_upgrade"""
+    post_upgrade = request.node.get_closest_marker('post_upgrade')
+    if post_upgrade:
+        depend_test = post_upgrade.kwargs.get('depend_on')
+        if depend_test:
+            pre_test_name = depend_test.__name__
+            # settings.configure()
+            cache_proxy = robottelo.config.settings_proxy._cache
+            # Read the worker id of pre test name
+            worker_int = int(get_worker_id(pre_test_name).replace('gw', ''))
+            cache_proxy['server.hostname'] = settings.server.hostnames[worker_int]
```

Comment:
There are some massive assumptions made in this implementation that concern me.

You need to align the test cases with the satellite target where depend_test actually ran - not the worker that it ran on.  If XDIST_BEHAVIOR was `balance` in pre_upgrade testing, or if any worker restarted and fetched a new on-demand satellite, then this mapping will be off.

The fixture implementing on-demand behavior to align xdist workers also runs in parallel on multiple threads, and its possible that worker `gw8` doesn't execute against the host from `server.hostnames` at index #8.

In fact, `server.hostnames` is only populated automatically when an inventory filter is supplied and a broker inventory is read.

Reply:
@jyejare let's discuss more tomorrow.  I think a fundamental change here would just be to store the hostname in pre_upgrade tests, instead of storing the worker ID. Then in post_upgrade, resolve the target hostname from the `depend_test`, instead of resolving a worker ID and then trying to infer a satellite hostname from that worker ID.

Reply:
Made the changes as expected.